### ISSUE TITLE
server: add functionality to limit amount of memory usage

### ISSUE THREAD

**Issue by niklasad1** [paritytech/jsonrpsee#1010]
> Very roughly maybe something like this could be better? (this proposal might still have some holes and need some tweaks/modifications, as I'm coming up with this on the spot)

> After serializing the reply into JSON before pushing it into MethodSink::tx calculate its size, and use that as the cost of the message.
    Have two per-connection limits - a soft memory usage limit, and a hard memory usage limit. If pending_messages_size + new_message_size > hard_limit then just force disconnect the client. If pending_messages_size + new_message_size > soft_limit then apply backpressure.
    Have a global memory usage limit across all connections. If that limit is breached find the biggest memory hog that is over the soft_limit and forcibly disconnect it. Repeat if we're still over the limit.
    If there are no more clients over the soft limit and we're still over the global memory usage limit then unconditionally apply backpressure, and reject any new subscriptions and requests with a "resources exceeded" or something like that (and on our load balancer we could get it to try load balance to another RPC server then or something like that)
    Also track the in-flight memory usage for non-subscription requests. (e.g. needed for the metadata endpoint)
    Also have a per-connection timeout that would kick in when a connection spends too much time over the soft limit and forcibly disconnect it (needed for our blocking handlers, some of which read from channels themselves; otherwise we'll just move the memory leak from jsonrpsee into those handler-side channels, and yes, we could "fix" those, but even so I'd rather have a plan B)
    (optional) Split big JSON replies into multiple chunks and deallocate them as they're sent so that we can gradually give back memory as the data is being sent.

> So this would achieve:

 >   The total memory usage of the RPC server would be roughly bounded.
 >  The most abusive/slowest clients would get punished first.
 >   Least abusive/fast clients under the limit would work as normal.
    The punishments are gradual based on memory usage - first throttling, then disconnection.
    If there's no one to punish/kick the server would gracefully decline new requests.
    Works both for subscriptions and non-subscriptions.

To summarize we would want the the total memory usage to be bounded roughly by the following:

- Rather than limit the number of messages (as currently), limit by total number of memory usage
- Add a custom websocket middleware trait to allow folks to write custom logic to punish clients, disconnect and similar.
